### DISCUSSION TITLE
Can't unit-test middleware

### DISCUSSION THREAD

**Discussion by jamesarosen** [vercel/next.js#61859] [Help]
### Summary

I'd like to write unit tests for my middleware. I'm using Jest and some of the problems below are specific to Jest, but others are common to any testing framework.

### Problem 1: No `Request`, `Response`, `fetch`

The [Next.js Jest docs](https://nextjs.org/docs/app/building-your-application/testing/jest) suggest using [jest-environment-jsdom](https://www.npmjs.com/package/jest-environment-jsdom), but jsdom hasn't addressed massive holes in their DOM support (e.g. `Request`, `Response`, `Headers`, `SubtleCrypto`, `AbortController`) [since 2017](https://github.com/jsdom/jsdom/issues/1724). Most of these can be relatively easily polyfilled if using Node ≥ 18.

### Problem 2: No `FetchEvent`

This is harder to polyfill because Node doesn't define `FetchEvent`.

### Problem 3: No `NextFetchEvent`

I can `import { NextFetchEvent } from 'next/server'`, but it's only a TypeScript type, not a function. I can't construct an instance of `NextFetchEvent` to pass to my middleware. Even worse: `NextFetchEvent` defines some of its prototype properties using `Symbol`s that it doesn't export, which means I can't even implement a `MockNextFetchEvent` that `implements NextFetchEvent`.

Building on this, there's no way to manually control the `waitUntil` queue in a test. That means I can't test that my middleware _doesn't_ do something during the request flow, but _does_ do it before hanging up.

**Comment by icyJoseph** [comment]
Hi,

For problem 1, there are two solutions:

Jest has something called `projects`, where you can define a different test environment for a given set of files. 

- https://jestjs.io/docs/configuration#projects-arraystring--projectconfig

The docs don't show it, but you can totally pass a `testEnvironment` entry and you'll be fine.

You can also do a one of environment setup with a magic comment:

```js
/**
 * @jest-environment node
 */

test('use Node.js in this test file', () => {
  
});
```

If you have Node 18+, then you'd have no trouble working with the Fetch family of APIs.

Regarding problem 2 and 3? I am not sure I follow? You should be able to do:

```ts
import { NextRequest, NextResponse, NextFetchEvent } from "next/server";

const fetchEvent = new NextFetchEvent({
  page: "/",
  request: new NextRequest(new Request("/")),
});
```

Or the likes. I think though, https://github.com/mockito/mockito is probably the way to go here, so that you mock only the properties you read, and then test the middleware response behavior.

**Reply by jamesarosen** [reply]
`new Request("/")` needs to be `new Request("https://example.com")` or similar because it has to be a full URL, but putting that aside...

`new NextFetchEvent(...)` causes the following exception:

```
TypeError: _server.NextFetchEvent is not a constructor
```

That's because [`next/server` doesn't export `NextFetchEvent`](https://github.com/vercel/next.js/blob/canary/packages/next/server.js). It's [only exported as a type](https://github.com/vercel/next.js/blob/895e7df86040a78964c86d72df58b6c2554e36ae/packages/next/server.d.ts#L7).

**Comment by Stiffni** [comment]
Having the same issue with `TypeError: NextFetchEvent is not a constructor`, did you end up finding a workaround?

**Comment by kjugi** [comment]
you can import from the direct path:

```js
import { NextFetchEvent } from 'next/dist/server/web/spec-extension/fetch-event';
```

at least in the next 14. Which should solve `...a constructor` TypeError

**Comment by kellyrmilligan** [comment]
is there an example somewhere of testing nextjs middleware?

**Reply by kjugi** [reply]
I have found at least 3 tests as examples:
- https://github.com/vercel/next.js/blob/89a72a05254aef1bd29f9ccb9dbdddf412b7f92f/test/e2e/app-dir/app-middleware/app-middleware.test.ts
- https://github.com/vercel/next.js/blob/89a72a05254aef1bd29f9ccb9dbdddf412b7f92f/test/e2e/middleware-general/test/index.test.ts
- https://github.com/vercel/next.js/blob/89a72a05254aef1bd29f9ccb9dbdddf412b7f92f/test/e2e/middleware-custom-matchers/test/index.test.ts

In your project, middleware itself can be tested like any other function. 

Pass the mocked variable with the `NextRequest` type and your data. Add `jest.fn()` in the places where your middleware is doing something (i.e. in for `cookie` object keys) so you can do`expect` and verify the final `NextResponse` return.

Personally, I would totally skip testing `config.matcher`. I assume it requires spawning the test `next` app and mocking upcoming requests but let's hear others 💭
